odl/discr/lp_discr.py

```diff
@@ -228,6 +234,12 @@ def exponent(self):
         """The exponent ``p`` in ``L^p``."""
         return self._exponent
 
+    @property
+    def vector_field_space(self):
+        """vector field space."""
+        self.real_space = self.astype(self._real_dtype)
+        return ProductSpace(self.real_space, self.ndim)
```

Comment:
I think it's rather straightforward - the gradient is an element of `X^d`, the Hessian (second derivative) in `X^(d x d)`, ..., the k-th derivative in `X^(d x ... x d)` with `k` factors.
Of course, another option `order` would be required...

Naming wouldn't be that horrible, like `tensor_field_space`.

A compromise could be to keep the property above and change the implementation to `return self.tensor_field_space(rank=1)` later.